src/logic/relation.lean

```diff
@@ -70,6 +70,28 @@ protected def map (r : α → β → Prop) (f : α → γ) (g : β → δ) : γ
 
 variables {r : α → α → Prop} {a b c d : α}
 
+/-- If a reflexive relation `r : α → α → Prop` holds over `x y : α`,
+then it holds even if `x ≠ y`. -/
+lemma reflexive_iff_ne_imp [is_refl α r] {x y : α} :
+  r x y ↔ (x ≠ y → r x y) :=
+begin
+  split;
+  intros h,
+  { intro,
+    apply h },
+  { by_cases hxy : x = y,
+    { rw hxy,
+      exact is_refl.refl _ },
+    { apply h hxy } }
+end
+
+/-- If a reflexive relation `r : α → α → Prop` holds over `x y : α`,
+then it holds even if `x ≠ y`. Unlike `reflexive_iff_ne_imp`, uses an explicit `reflexive r`
+instead of `[is_refl α r]`.  -/
+lemma reflexive.iff_ne_imp (h : reflexive r) {x y : α} :
+  r x y ↔ (x ≠ y → r x y) :=
+@reflexive_iff_ne_imp _ r ⟨h⟩ _ _
```

Comment:
Good idea adding this lemma. I think it would make more sense to prove them in the opposite direction though, and prove `reflexive_iff_ne_imp` in terms of `reflexive.iff_ne_imp`